plugins/catalog-backend/src/catalog/DatabaseCatalog.ts

```diff
@@ -44,15 +45,27 @@ export class DatabaseCatalog implements Catalog {
     private readonly logger: Logger,
   ) {}
 
-  async addOrUpdateComponent(component: Component): Promise<Component> {
+  async addOrUpdateComponent(
+    locationId: string,
+    descriptor: ComponentDescriptor,
+  ): Promise<void> {
+    const component: Component = {
+      name: descriptor.metadata.name,
+    };
+
     await this.database.transaction(async (tx) => {
-      await tx('components')
-        .insert(component)
-        .catch(() =>
-          tx('components').where({ name: component.name }).update(component),
-        );
+      // TODO(freben): Currently, several locations can compete for the same component
+      const count = await tx('components')
+        .where({ name: component.name })
+        .update({ ...component, locationId });
```

Comment:
I'm assuming the order, first `update` then `insert` doesn't really matter? Or is one more common then the other?

Reply:
Either could work, but since updates will be way more common than inserts, it's better to treat that as the default case.

Plus, if I do insert first, then it's an exception (key violation) that one handles, so it felt less violent to just try to update first and then insert if the first thing did nothing :)